### ISSUE TITLE
Promise can be resolved with undefined but not reflected in type

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#11094]
TS 2.0.3

``` ts
    new Promise<string>(resolve => {
        true ? resolve('foo') :  resolve()
    })
```

This returns a `Promise<string>`, however it is possible to call `resolve` with no value, meaning the return type is actually `Promise<string | undefined>` at runtime.

I'm not sure whether the return type should in fact be inferred as `Promise<string | undefined>` or whether the empty `resolve()` call should throw a compile error.

**Comment by kitsonk** [comment]
Dupe of #3356

**Comment by OliverJAsh** [comment]
#3356 appears to be marked as a dupe of https://github.com/Microsoft/TypeScript/issues/4260.

What's the status of this now that we have strict null checking (since #3356)?

**Comment by MicahZoltu** [comment]
Pinging on the status of this.  At the moment, with strict null checking the [Promise definition file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/es6-promise/index.d.ts#L19) is simply incorrect.

```
constructor(callback: (resolve : (value?: T | Thenable<T>) => void, reject: (error?: any) => void) => void);
```
Notice how `resolve` is a function that takes in an optional `T` or `Thenable<T>`.  This isn't correct.  It should take in a required `T` or `Thenable<T>`.  It is a programming error to resolve a promise with the wrong type.  If you want a promise of an optional that is fine, but your `T` should be `T|undefined` in this case.

There are a number of copies of this issue floating around and most of them are labeled as a dupe of https://github.com/Microsoft/TypeScript/issues/4260, however on cursory glance of that issue I think perhaps the reality is that #4260 is a a prerequisite before fixing this issue.  Once 4260 is fixed then the definition for Promise can be fixed to work correctly with strict null checking.
